src/Microsoft.Health.Fhir.Shared.Api/Controllers/ValidateController.cs

```diff
@@ -50,26 +51,48 @@ public ValidateController(IMediator mediator, IOptions<FeatureConfiguration> fea
         [Route(KnownRoutes.ValidateResourceType)]
         [AuditEventType(AuditEventSubType.Read)]
         [Authorize(PolicyNames.ReadPolicy)]
-        public async Task<IActionResult> Validate([FromBody] Resource resource, [FromQuery(Name = KnownQueryParameterNames.Profile)] string profile, [FromQuery(Name = KnownQueryParameterNames.Mode)] string mode)
+        public async Task<IActionResult> Validate([FromBody] Resource resource, [FromQuery(Name = KnownQueryParameterNames.Profile)] string profile, [FromQuery(Name = KnownQueryParameterNames.Mode)] string mode, string typeParameter)
+        {
+            return await RunValidationAsync(resource, profile, mode, typeParameter);
+        }
+
+        [HttpPost]
+        [Route(KnownRoutes.ValidateResourceTypeById)]
+        [AuditEventType(AuditEventSubType.Read)]
+        [Authorize(PolicyNames.ReadPolicy)]
+        public async Task<IActionResult> ValidateById([FromBody] Resource resource, [FromQuery(Name = KnownQueryParameterNames.Profile)] string profile, [FromQuery(Name = KnownQueryParameterNames.Mode)] string mode, string typeParameter, string idParameter)
+        {
+            return await RunValidationAsync(resource, profile, mode, typeParameter, idParameter, true);
+        }
+
+        private async Task<IActionResult> RunValidationAsync(Resource resource, string profile, string mode, string typeParameter, string idParameter = null, bool idMode = false)
         {
             if (!_features.SupportsValidate)
             {
                 throw new OperationNotImplementedException(Resources.ValidationNotSupported);
             }
 
-            if (resource.ResourceType == ResourceType.Parameters)
+            if (profile != null)
             {
-                throw new OperationNotImplementedException(Resources.ValidateWithParametersNotSupported);
+                throw new OperationNotImplementedException(Resources.ValidateWithProfileNotSupported);
             }
 
-            if (!string.IsNullOrEmpty(profile))
+            if (resource.ResourceType == ResourceType.Parameters)
             {
-                throw new OperationNotImplementedException(Resources.ValidateWithProfileNotSupported);
+                resource = ParseParameters((Parameters)resource, ref profile, ref mode);
             }
 
-            if (!string.IsNullOrEmpty(mode))
+            // This is the same as the filter that is applied in the ValidationModeFilter.
+            // It is needed here to cover the case of the mode being passed as part of a Parameters resource.
```

Comment:
Would it be possible to parse the resource in the `ValidationModeFilter`, check if it is `Parameters` type and do this validation in the filter itself?

Reply:
The issue with doing it in the filter is that it is then done for all of our endpoints, as they all share the validate filter. Since the Parameters syntax is endpoint specific (different endpoints can have different properties) doing this would mean this controller would need a filter specifically for it.

Reply:
Ignore my above comment. I refactored the mode filter to do all the work.